### PR TITLE
TST: Reproduce paintEvent failures

### PR DESCRIPTION
**Pull request by GuillaumeFavelier**
The goal is to fix the Azure failures encountered recently during testing in: https://github.com/mne-tools/mne-python/pull/7290, https://github.com/mne-tools/mne-python/issues/7235

The log is very scarce, I'm trying locally with `virtualbox` at the same time:
```
mne\viz\_3d.py s                                                         [ 91%]
mne\viz\circle.py s                                                      [ 91%]
mne\viz\misc.py .                                                        [ 91%]
mne\viz\utils.py s                                                       [ 91%]
mne\viz\_brain\tests\test_brain.py ......ssssssss.                       [ 92%]
mne\viz\backends\tests\test_renderer.py .sx..ss                          [ 92%]
mne\viz\backends\tests\test_utils.py ..                                  [ 93%]
Windows fatal exception: access violation

Thread 0x00000ca8 (most recent call first):
  File "d:\a\1\s\conda\lib\threading.py", line 296 in wait
  File "d:\a\1\s\conda\lib\queue.py", line 170 in get
  File "D:\a\1\s\mne\gui\_coreg_gui.py", line 1363 in worker
  File "d:\a\1\s\conda\lib\threading.py", line 870 in run
  File "d:\a\1\s\condaWindows fatal exception: \access violationl

ib\threading.py", line 926 in _bootstrap_inner
##[error]Cmd.exe exited with code '-1073741819'.
```

### CODE REVIEW

**Comment by GuillaumeFavelier** [comment]
Closing now since this seems to be solved by https://github.com/mne-tools/mne-python/pull/7295

**Comment by GuillaumeFavelier** [comment]
The failures still occur.

Reference:
https://github.com/mne-tools/mne-python/pull/7342#issuecomment-588534058

**Comment by larsoner** [comment]
Here is a nice readable traceback

https://dev.azure.com/mne-tools/mne-python/_build/results?buildId=5624&view=logs&jobId=20ec0733-a031-5a42-425a-ca7f9f76ad75&j=20ec0733-a031-5a42-425a-ca7f9f76ad75&t=3732c0a7-07d3-577b-8168-001d94589ebf

**Comment by GuillaumeFavelier** [comment]
I was able to read it when you did the comment but I forgot to save the log, now the build is not available anymore. I remember it was related to `QVTKRenderWindowInteractor` and `PainEvent`.

**Comment by larsoner** [comment]
Just go through the recent Azure builds, it usually only takes looking at a few to find a failure with a good traceback, like:

https://dev.azure.com/mne-tools/mne-python/_build/results?buildId=5634&view=logs&j=b195a6e3-7484-5943-e627-0b35c27605d6&t=6dcb3581-08d4-5e76-74f2-6edaf20ff912&l=257

**Comment by GuillaumeFavelier** [comment]
For reference, here is a log:

<details>

```
mne\utils\tests\test_testing.py ....                                     [ 91%]
mne\viz\_3d.py s                                                         [ 91%]
mne\viz\circle.py s                                                      [ 91%]
mne\viz\misc.py .                                                        [ 91%]
mne\viz\utils.py s                                                       [ 92%]
mne\viz\_brain\tests\test_brain.py ..............                        [ 92%]
mne\viz\backends\tests\test_renderer.py ..x....                          [ 93%]
mne\viz\backends\tests\test_utils.py ..                                  [ 93%]
Windows fatal exception: access violation

Thread 0x00001198 (most recent call first):
  File "d:\a\1\s\conda\lib\site-packages\tvtk\pyface\ui\qt4\QVTKRenderWindowInteractor.py", line 370 in paintEvent
  File "d:\a\1\s\conda\lib\site-packages\tvtk\pyface\ui\qt4\scene.py", line 69 in paintEvent
  File "d:\a\1\s\conda\lib\site-packages\pyface\ui\qt4\gui.py", line 85 in process_events
  File "d:\a\1\s\conda\lib\site-packages\surfer\viz.py", line 183 in _force_render
  File "d:\a\1\s\conda\lib\site-packages\surfer\viz.py", line 550 in _toggle_render
  File "d:\a\1\s\conda\lib\site-packages\surfer\viz.py", line 487 in __init__
  File "D:\a\1\s\mne\viz\_3d.py", line 2385 in plot_vector_source_estimates
  File "D:\a\1\s\mne\utils\_logging.py", line 89 in wrapper
  File "<D:\a\1\s\mne\externals\decorator.py:decorator-gen-131>", line 2 in plot_vector_source_estimates
  File "D:\a\1\s\mne\source_estimate.py", line 2197 in plot
  File "D:\a\1\s\mne\viz\tests\test_3d.py", line 651 in test_plot_vector_source_estimates
  File "D:\a\1\s\mne\utils\_testing.py", line 259 in dec
  File "d:\a\1\s\conda\lib\site-packages\_pytest\python.py", line 167 in pytest_pyfunc_call
  File "d:\a\1\s\conda\lib\site-packages\pluggy\callers.py", line 187 in _multicall
  File "d:\a\1\s\conda\lib\site-packages\pluggy\manager.py", line 87 in <lambda>
  File "d:\a\1\s\conda\lib\site-packages\pluggy\manager.py", line 93 in _hookexec
  File "d:\a\1\s\conda\lib\site-packages\pluggy\hooks.py", line 286 in __call__
  File "d:\a\1\s\conda\lib\site-packages\_pytest\python.py", line 1445 in runtest
  File "d:\a\1\s\conda\lib\site-packages\_pytest\runner.py", line 134 in pytest_runtest_call
  File "d:\a\1\s\conda\lib\site-packages\pluggy\callers.py", line 187 in _multicall
  File "d:\a\1\s\conda\lib\site-packages\pluggy\manager.py", line 87 in <lambda>
  File "d:\a\1\s\conda\lib\site-packages\pluggy\manager.py", line 93 in _hookexec
  File "d:\a\1\s\conda\lib\site-packages\pluggy\hooks.py", line 286 in __call__
  File "d:\a\1\s\conda\lib\site-packages\_pytest\runner.py", line 210 in <lambda>
  File "d:\a\1\s\conda\lib\site-packages\_pytest\runner.py", line 237 in from_call
  File "d:\a\1\s\conda\lib\site-packages\_pytest\runner.py", line 210 in call_runtest_hook
  File "d:\a\1\s\conda\lib\site-packages\_pytest\runner.py", line 185 in call_and_report
  File "d:\a\1\s\conda\lib\site-packages\_pytest\runner.py", line 99 in runtestprotocol
  File "d:\a\1\s\conda\lib\site-packages\_pytest\runner.py", line 84 in pytest_runtest_protocol
  File "d:\a\1\s\conda\lib\site-packages\pluggy\callers.py", line 187 in _multicall
  File "d:\a\1\s\conda\lib\site-packages\pluggy\manager.py", line 87 in <lambda>
  File "d:\a\1\s\conda\lib\site-packages\pluggy\manager.py", line 93 in _hookexec
  File "d:\a\1\s\conda\lib\site-packages\pluggy\hooks.py", line 286 in __call__
  File "d:\a\1\s\conda\lib\site-packages\_pytest\main.py", line 271 in pytest_runtestloop
  File "d:\a\1\s\conda\lib\site-packages\pluggy\callers.py", line 187 in _multicall
  File "d:\a\1\s\conda\lib\site-packages\pluggy\manager.py", line 87 in <lambda>
  File "d:\a\1\s\conda\lib\site-packages\pluggy\manager.py", line 93 in _hookexec
  File "d:\a\1\s\conda\lib\site-packages\pluggy\hooks.py", line 286 in __call__
  File "d:\a\1\s\conda\lib\site-packages\_pytest\main.py", line 247 in _main
  File "d:\a\1\s\conda\lib\site-packages\_pytest\main.py", line 197 in wrap_session
  File "d:\a\1\s\conda\lib\site-packages\_pytest\main.py", line 240 in pytest_cmdline_main
  File "d:\a\1\s\conda\lib\site-packages\pluggy\callers.py", line 187 in _multicall
  File "d:\a\1\s\conda\lib\site-packages\pluggy\manager.py", line 87 in <lambda>
  File "d:\a\1\s\conda\lib\site-packages\pluggy\manager.py", line 93 in _hookexec
  File "d:\a\1\s\conda\lib\site-packages\pluggy\hooks.py", line 286 in __call__
  File "d:\a\1\s\conda\lib\site-packages\_pytest\config\__init__.py", line 93 in main
  File "d:\a\1\s\conda\Scripts\pytest-script.py", line 10 in <module>
```

</details>

**Comment by larsoner** [comment]
Travis even hit one:

https://travis-ci.org/mne-tools/mne-python/jobs/655188026#L2938-L2944

**Comment by GuillaumeFavelier** [comment]
The latest commit https://github.com/mne-tools/mne-python/pull/7380/commits/499f917dd301826606f706e775c611c3e215d1b0 on #7380 seem to reproduce this behaviour on Travis consistently.

**Comment by GuillaumeFavelier** [comment]
Reproduced in here, it appears to be consistent.

**Comment by GuillaumeFavelier** [comment]
My first idea is to use the `renderer` fixture in `test_mixed_sources_plot_surface`

**Comment by GuillaumeFavelier** [comment]
Test 1) is green

File: mne/viz/tests/test_3d.py [hunk 1]
```diff
@@ -719,7 +715,7 @@ def test_brain_colorbar(orientation, diverging, lims):
 @requires_pysurfer
 @testing.requires_testing_data
 @traits_test
-def test_mixed_sources_plot_surface(garbage_collect):
+def test_mixed_sources_plot_surface(garbage_collect, renderer):
```
Comment (larsoner) [inline mne/viz/tests/test_3d.py]: You can get rid of the garbage collect, it's redundant once renderer is added

**Comment by GuillaumeFavelier** [comment]
I ported 532aba9 in https://github.com/mne-tools/mne-python/pull/7380

**Comment by GuillaumeFavelier** [comment]
Travis jobs are up in here, waiting for the good news :crossed_fingers:

**Comment by agramfort** [comment]
thx @GuillaumeFavelier !

**Comment by agramfort** [comment]
thx @GuillaumeFavelier !

File: mne/viz/_brain/_brain.py [hunk 2]
```diff
@@ -819,71 +819,39 @@ def set_data_smoothing(self, n_steps):
...
-        act_data = self._time_interp_funcs[hemi](time_idx)
-        act_time = self._time_interp_inv(time_idx)
+    def _interpolate_data(self, array, time_idx):
+        from scipy.interpolate import interp1d
+        times = np.arange(self._n_times)
+        act_data = interp1d(
+            times, array, self.interp_kind, axis=1,
+            assume_sorted=True)(time_idx)
+        ifunc = interp1d(times, self._data['time'])
+        act_time = ifunc(time_idx)
```
Comment (larsoner) [inline mne/viz/_brain/_brain.py]: This PR undid (again) some of the refactoring. The way my commits worked:

1. the interpolators were [set up once and reused](https://github.com/mne-tools/mne-python/pull/7257/commits/be72e644405bb5a0ece8b04fc8059730f76a22dc) (efficiency)
2. a helper [_safe_interp1d](https://github.com/mne-tools/mne-python/pull/7257/commits/da82289a1f031f7a28120caade5c5bb598b7897e) (avoids bugs now back in master)

And this commit appears to have un-done these. Accordingly we have lost the efficiency and the singleton regression is back in `master`:

https://circleci.com/gh/mne-tools/mne-python/18710?utm_campaign=workflow-failed&utm_medium=email&utm_source=notification


I think the real (only) bug that probably needed to be fixed here had to do with calling the interpolation functions incorrectly, right? If so @GuillaumeFavelier can you make a revert commit for this PR and try making the minimal changes required to get CIs happy again? I'd rather not sacrifice the `_safe_interp1d` and `funcs` efficiency gains...

Comment (larsoner) [reply mne/viz/_brain/_brain.py]: Ahh it's even more extensive -- I see this started out by completely reverting #7257. Are we planning on putting that back in somehow?

Comment (GuillaumeFavelier) [reply mne/viz/_brain/_brain.py]: I updated #7162 and delayed thoses features to 0.21 for now

**Pull request merged**